File: src/java/src/DisjointTypes.java
//: src/java/src/DisjointTypes.java

class Dog {
  void gulp() {
    System.out.println("eating dog food");
  }
  void bark() {
    System.out.println("woof");
  }
}

class Person {
  void scarf() {
    System.out.println("eating pizza");
  }
  void greet() {
    System.out.println("hello");
  }
}
//---
class Robot {
  void charge() {
    System.out.println("charging");
  }
  void initiate() {
    System.out.println("operational");
  }
}

class Slug {
  void absorb() {
    System.out.println("eating grass");
  }
}
//---
public class DisjointTypes {
  static void nourish(Dog subject) {
    subject.gulp();
    subject.bark();
  }
  static void nourish(Person subject) {
    subject.scarf();
    subject.greet();
  }
  static void nourish(Robot subject) {
    subject.charge();
    subject.initiate();
  }
//---
  static void nourish(Slug subject) {
    subject.absorb();
  }
  public static void main(String... args) {
    // Ad-hoc polymorphism (overloading):
    nourish(new Dog());
    nourish(new Person());
    nourish(new Robot());
    nourish(new Slug());
    // (Statically determined at compile time)
  }
}

File: src/python/disjoint_types.py
#: src/python/disjoint_types.py
class Dog:
    def eat(self): print("eating dog food")
    def bark(self): print("woof")

class Person:
    def eat(self): print("eating pizza")
    def greet(self): print("hello")

class Robot:
    def eat(self): print("charging")
    def initiate(self): print("operational")

class Slug:
    def eat(self): print("eating grass")

def nourish(subject):
    subject.eat()  # Duck typing
    if type(subject) is Dog: subject.bark()
    if type(subject) is Person: subject.greet()
    if type(subject) is Robot: subject.initiate()

# Sumtype polymorphism:
def nourish2(subject: Dog | Person | Robot | Slug):
    subject.eat()
    match subject:
        case Dog(): subject.bark()
        case Person(): subject.greet()
        case Robot(): subject.initiate()

if __name__ == '__main__':
    for subject in [Dog(), Person(), Robot(), Slug()]:
        nourish(subject)
        nourish2(subject)
        # nourish2("")
